ITE/10/README.adoc

```diff
@@ -0,0 +1,278 @@
+assumption that a step in the software supply chain affects in some manner the
+artifacts that constitute the supply chain. ITE-6 showed that some steps do not
+fit the materials and products model. This is a factor in _verifying_ such
+predicates using in-toto layouts.
+
+Therefore, all predicates supported for verification in an in-toto layout must
+categorized as one of two classes:
+
+* transformational predicates: these are based on classic in-toto links for
+  steps that transform artifacts in some manner. They consume some artifacts as
+  materials and produce others as products. The subject of such attestations
+  indicate the products of the step. Eg. in-toto links, SLSA Provenance.
+* informational predicates: these predicates record contextual attributes about
+  one or more artifacts. The operations they represent consume artifacts as
+  materials (recorded as the subject of such attestations) and produce
+  attributes about them as the outcome. Eg. test results, reviews.
+
+Note that transformational predicates may also record certain information about
+certain attributes. For example, the environment field in in-toto links can
+include such information.
+
+With this representation of different predicate types, step declarations in
+layouts can _chain_ different operations represented by distinct predicate types
+together, a fundamental feature of in-toto. For example, a code review step
+represented by a review predicate type that follows a checkout step that fetches
+source code represented by an in-toto link can use in-toto's artifact rules to
+ensure the review was performed against the sources fetched by the checkout
+step.
+
+=== Setting Predicate Type Expectations for Steps and Inspections
+
+Currently, step declarations have the following schema.
+
+```yaml
+name: string
+threshold: int
+expectedMaterials: list of artifact rules
+expectedProducts: list of artifact rules
+pubkeys: list of authorized key IDs
+expectedCommand: list of strings
+```
+
+This ITE proposes updating this schema to incorporate predicate types.
+
+```yaml
+name: string
+command: list of strings
+expectedMaterials: list of artifact rules
+expectedPredicates: list of expectedStepPredicates
+```
+
+Each `expectedStepPredicate` object has the following schema.
+
+```yaml
+predicateType: string
+expectedProducts: list of artifact rules
+functionaries: list of authorized functionary key IDs
+threshold: int
+```
+
+The `predicateType` field identifies the type of predicate expected for the
+step. `expectedProducts` is moved to being predicate specific as informational
+predicates may be generated alongside transformational predicates for the same
+step. This means that product rules do not generalize across all of the step's
+predicates. However, `expectedMaterials` is retained at the step level as all
+predicates generated for a step are expected to be for the same set of
+artifacts. `pubkeys` is updated to `functionaries` to generalize signing
```

Comment:
Maybe I'm missing something, but does leaving `expectedMaterials` at the step level mean that verifiers will expect to informational predicates to indicate the set of materials they apply to? Or is the assumption here that informational predicates will always indicate the `expectedMaterials` in the attestation subject field? 

I'm imagining a use case where my step generates some informational predicates about things like the tool or runtime system that ran during the step, such as the compiler or platform. Does the compiler that I ran during the step become an `expectedMaterial`? I think this also relates to my earlier comment about specifying the `expectedSubjects` in some cases rather than using the transformation-oriented terms "materials" and "products".

Reply:
Yes, informational attestations will always list materials in their subject fields. However, this is not directly linked to the `expectedMaterials` being at the step level. That's to indicate that if a step can generate more than one attestation (of different predicate types, so not handled by the threshold), they must all have the same materials. Generally, it formalizes that two attestations with different predicate types must have the same materials to be considered part of the same step. Note that this does not mean that all attestations with the same materials must belong to a single step. I took this approach because not all predicates have products but they all do have materials.

>I'm imagining a use case where my step generates some informational predicates about things like the tool or runtime system that ran during the step, such as the compiler or platform. Does the compiler that I ran during the step become an expectedMaterial?

I think not, actually. The step generates this information about some artifact, so the attestation's subject is the artifact, right? The compiler info is within the predicate itself. So the attestation can be thought of representing a "tool-info" step?